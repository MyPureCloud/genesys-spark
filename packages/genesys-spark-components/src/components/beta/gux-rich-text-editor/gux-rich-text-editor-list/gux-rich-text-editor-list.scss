@use '~genesys-spark/dist/beta/tokens/flare/ui.scss';

:host {
  display: flex;
  flex-direction: column;
  flex-wrap: nowrap;
  place-content: stretch flex-start;
  align-items: flex-start;
  padding: ui.$gse-ui-menu-padding;
}
