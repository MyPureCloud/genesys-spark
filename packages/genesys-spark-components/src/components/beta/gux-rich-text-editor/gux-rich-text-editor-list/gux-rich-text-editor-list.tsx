import {
  Component,
  Element,
  Listen,
  Host,
  h,
  Method,
  Prop,
  State,
  Event,
  EventEmitter
} from '@stencil/core';
import { trackComponent } from '@utils/tracking/usage';

import {
  first,
  last,
  next,
  previous
} from '../../../stable/gux-list/gux-list.service';

/**
 * @slot - collection of gux-rich-style-list-item, gux-rich-highlight-list-item or gux-rich-text-editor-sub-list elements.
 */

const validFocusableItems = [
  'gux-rich-style-list-item',
  'gux-rich-highlight-list-item',
  'gux-rich-text-editor-sub-list'
];

@Component({
  tag: 'gux-rich-text-editor-list',
  styleUrl: 'gux-rich-text-editor-list.scss',
  shadow: { delegatesFocus: true }
})
export class GuxRichTextEditorList {
  @Element()
  root: HTMLElement;

  @Prop()
  value: string;

  @State()
  listItems: HTMLElement[] = [];

  @Event()
  internallistitemsupdated: EventEmitter;

  get listItemsSlot(): HTMLSlotElement | null {
    return this.root.querySelector('slot');
  }

  get listItemElements(): HTMLElement[] {
    const assignedElements = this.listItemsSlot?.assignedElements();

    if (assignedElements) {
      return Array.from(assignedElements as HTMLElement[]);
    }

    return [];
  }

  //Set the first gux-rich-highlight-list-item tab-index to 0 so we can tab to the list of colors.
  private handleHighlighItemsNavigation(): void {
    if (this.listItemsSlot) {
      const firstHighlightItem = this.listItemsSlot
        .assignedElements()
        .find(
          el => el.tagName.toLowerCase() === 'gux-rich-highlight-list-item'
        );
      if (firstHighlightItem) {
        const buttonElement =
          firstHighlightItem.shadowRoot.querySelector('button');
        buttonElement.tabIndex = 0;
      }
    }
  }

  private setListItems(): void {
    this.listItems = this.listItemElements as HTMLElement[];
    this.internallistitemsupdated.emit();
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root);

    this.setListItems();
    this.handleHighlighItemsNavigation();
  }

  @Listen('keydown')
  onKeyDown(event: KeyboardEvent): void {
    const target = event.target as HTMLElement;

    if (!target) {
      return;
    }

    const keyHandlers = {
      ArrowUp: () => previous(this.root, validFocusableItems),
      ArrowDown: () => next(this.root, validFocusableItems),
      Home: () => first(this.root, validFocusableItems),
      End: () => last(this.root, validFocusableItems)
    };

    if (target.matches('gux-rich-highlight-list-item')) {
      Object.assign(keyHandlers, {
        ArrowLeft: () => previous(this.root, validFocusableItems),
        ArrowRight: () => next(this.root, validFocusableItems)
      });
    }

    const handler = keyHandlers[event.key];
    if (handler) {
      event.preventDefault();
      handler();
    }
  }

  @Method()
  // eslint-disable-next-line @typescript-eslint/require-await
  async guxFocusFirstItem(): Promise<void> {
    first(this.root, validFocusableItems);
  }

  @Method()
  // eslint-disable-next-line @typescript-eslint/require-await
  async guxFocusLastItem(): Promise<void> {
    last(this.root, validFocusableItems);
  }

  render(): JSX.Element {
    return (
      <Host role="list">
        <slot onSlotchange={() => this.setListItems()}></slot>
      </Host>
    );
  }
}
