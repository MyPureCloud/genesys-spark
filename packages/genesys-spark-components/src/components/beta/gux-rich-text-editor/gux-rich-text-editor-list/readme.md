# gux-rich-text-editor-list



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description | Type     | Default     |
| -------- | --------- | ----------- | -------- | ----------- |
| `value`  | `value`   |             | `string` | `undefined` |


## Events

| Event                      | Description | Type               |
| -------------------------- | ----------- | ------------------ |
| `internallistitemsupdated` |             | `CustomEvent<any>` |


## Methods

### `guxFocusFirstItem() => Promise<void>`



#### Returns

Type: `Promise<void>`



### `guxFocusLastItem() => Promise<void>`



#### Returns

Type: `Promise<void>`




## Dependencies

### Used by

 - [gux-rich-text-editor-action-rich-style](../gux-rich-text-editor-action/gux-rich-text-editor-action-rich-style)

### Graph
```mermaid
graph TD;
  gux-rich-text-editor-action-rich-style --> gux-rich-text-editor-list
  style gux-rich-text-editor-list fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
