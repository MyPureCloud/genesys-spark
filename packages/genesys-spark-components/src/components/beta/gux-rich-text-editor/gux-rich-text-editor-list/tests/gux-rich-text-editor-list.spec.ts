import { newSpecPage } from '@test/specTestUtils';

import { GuxRichTextEditorList } from '../gux-rich-text-editor-list';

const components = [GuxRichTextEditorList];
const language = 'en';

describe('gux-rich-text-editor-list', () => {
  it('should build and render', async () => {
    const html = `
    <gux-rich-text-editor-list>
      <gux-rich-style-list-item value="heading-1"><h1>Heading 1</h1></gux-rich-style-list-item>
      <gux-rich-style-list-item value="heading-2"><h2>Heading 2</h2></gux-rich-style-list-item>
      <gux-rich-style-list-item value="heading-3"><h3>Heading 3</h3></gux-rich-style-list-item>
      <gux-rich-style-list-item value="paragraph"><p>Paragraph</p></gux-rich-style-list-item>
    </gux-rich-text-editor-list>
    `;
    const page = await newSpecPage({ components, html, language });

    expect(page.rootInstance).toBeInstanceOf(GuxRichTextEditorList);
    expect(page.root).toMatchSnapshot();
  });
});
