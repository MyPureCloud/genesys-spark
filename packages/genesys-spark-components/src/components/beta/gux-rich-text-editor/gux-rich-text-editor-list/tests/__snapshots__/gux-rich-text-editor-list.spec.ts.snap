// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-rich-text-editor-list should build and render 1`] = `
<gux-rich-text-editor-list role="list">
  <template shadowrootmode="open">
    <span aria-hidden="true" tabindex="-1"></span>
    <slot></slot>
  </template>
  <gux-rich-style-list-item value="heading-1">
    <h1>
      Heading 1
    </h1>
  </gux-rich-style-list-item>
  <gux-rich-style-list-item value="heading-2">
    <h2>
      Heading 2
    </h2>
  </gux-rich-style-list-item>
  <gux-rich-style-list-item value="heading-3">
    <h3>
      Heading 3
    </h3>
  </gux-rich-style-list-item>
  <gux-rich-style-list-item value="paragraph">
    <p>
      Paragraph
    </p>
  </gux-rich-style-list-item>
</gux-rich-text-editor-list>
`;
